File: backend/src/services/whatIfService.js
import { pathService } from "./pathService.js";

export class WhatIfService {
  executeScenario(request) {
    const profileData = request.student_profile || {};
    const student = {
      goal: profileData.goal || "AI Engineer",
      available_hours_per_week: parseFloat(profileData.available_hours_per_week || 6.0),
    };

    const basePaths = pathService.generatePathsForStudent(student);
    const originalPath = request.custom_base_path || basePaths[0];

    // Deep clone original path
    const altPath = JSON.parse(JSON.stringify(originalPath));
    const scenario = request.scenario || {};
    const changes = [];
    const tradeOffs = [];
    const op = (scenario.operation || "").toUpperCase();

    if (op === "REPLACE" || (scenario.operation || "").toLowerCase().includes("replace")) {
      const targetName = (scenario.target || "AI Club").toLowerCase();

      let removedStepName = "";
      const newSteps = [];
      for (const s of altPath.steps || []) {
        if (s.name.toLowerCase().includes(targetName) || s.type.toLowerCase().includes(targetName)) {
          removedStepName = s.name;
          changes.push(`Removed '${s.name}' (${s.hours_per_week}h/wk)`);
        } else {
          newSteps.push(s);
        }
      }

      if (!removedStepName && (altPath.steps || []).length >= 3) {
        removedStepName = altPath.steps[2].name;
        changes.push(`Removed '${removedStepName}' (${altPath.steps[2].hours_per_week}h/wk)`);
        newSteps.splice(2, 1);
      }

      const researchStep = {
        step_id: "step_whatif_res_1",
        type: "Research",
        name: "RES_01: Autonomous Navigation & Perception Research",
        description: "Faculty-led computer vision research with Prof. Aris Thorne.",
        hours_per_week: 1.5,
        duration: "1 Semester",
        skills: ["Computer Vision", "Deep Learning", "Research Methods", "Faculty Mentorship"],
        reason: "Replaces extracurricular club with direct faculty-led research exposure.",
        prerequisites: ["AI101", "Python"],
        dependencies: ["step_a_2"],
      };

      newSteps.push(researchStep);
      changes.push(`Added research alternative '${researchStep.name}' (${researchStep.hours_per_week}h/wk)`);

      altPath.steps = newSteps;
      altPath.title = "Path A' — What-If Alternative (Research Substituted)";
      altPath.focus_type = "Research Substituted";
      altPath.description = "Alternative candidate path substituting AI Club with faculty computer vision research.";

      // Update metrics
      altPath.metrics.research_exposure = Math.min(100.0, originalPath.metrics.research_exposure + 50.0);
      altPath.metrics.faculty_interaction = Math.min(100.0, originalPath.metrics.faculty_interaction + 45.0);
      altPath.metrics.networking_value = Math.max(10.0, originalPath.metrics.networking_value - 25.0);
      altPath.metrics.industry_exposure = Math.max(10.0, originalPath.metrics.industry_exposure - 15.0);
      altPath.metrics.goal_alignment = 95.0;

      tradeOffs.push("Gains direct faculty mentorship and lab publication potential.");
      tradeOffs.push("Reduces general peer networking and club social events.");
    } else {
      // General What-if adjustment
      changes.push(`Simulated scenario '${scenario.target || "custom modification"}'`);
      tradeOffs.push("Maintains target time budget while enhancing specialized skills.");
    }

    // Recalculate total hours
    altPath.total_hours_per_week = altPath.steps.reduce((acc, s) => acc + s.hours_per_week, 0);
    altPath.within_limit = altPath.total_hours_per_week <= altPath.available_hours_per_week;

    const metricComparisons = [
      {
        metric_name: "Research Exposure",
        original_value: originalPath.metrics.research_exposure,
        simulated_value: altPath.metrics.research_exposure,
        delta: altPath.metrics.research_exposure - originalPath.metrics.research_exposure,
        interpretation: "Significant increase in academic lab and publication experience.",
      },
      {
        metric_name: "Faculty Interaction",
        original_value: originalPath.metrics.faculty_interaction,
        simulated_value: altPath.metrics.faculty_interaction,
        delta: altPath.metrics.faculty_interaction - originalPath.metrics.faculty_interaction,
        interpretation: "Direct interaction with research faculty.",
      },
      {
        metric_name: "Networking & Clubs",
        original_value: originalPath.metrics.networking_value,
        simulated_value: altPath.metrics.networking_value,
        delta: altPath.metrics.networking_value - originalPath.metrics.networking_value,
        interpretation: "Slight decrease in broad peer networking.",
      },
      {
        metric_name: "Industry Readiness",
        original_value: originalPath.metrics.industry_exposure,
        simulated_value: altPath.metrics.industry_exposure,
        delta: altPath.metrics.industry_exposure - originalPath.metrics.industry_exposure,
        interpretation: "Balanced between industry tooling and research.",
      },
    ];

    return {
      original_path: originalPath,
      simulated_path: altPath,
      metric_comparisons: metricComparisons,
      changes_summary: changes,
      trade_offs: tradeOffs,
      recommendation:
        "Replacing club participation with faculty research substantially accelerates your academic profile while keeping total hours within your target budget.",
      time_impact: `${altPath.total_hours_per_week}h/wk total (${altPath.total_hours_per_week - originalPath.total_hours_per_week >= 0 ? "+" : ""}${Math.round((altPath.total_hours_per_week - originalPath.total_hours_per_week) * 10) / 10}h change)`,
    };
  }
}

export const whatIfService = new WhatIfService();

File: backend/app/services/what_if_service.py
import copy
from typing import List, Dict, Any, Optional
from app.models.student import StudentProfile
from app.models.path import CampusPath, PathStep, PathMetrics
from app.models.what_if import WhatIfRequest, WhatIfResponse, WhatIfScenario, MetricComparison
from app.services.path_service import path_service

class WhatIfService:
    def execute_scenario(self, request: WhatIfRequest) -> WhatIfResponse:
        # 1. Base profile
        profile_data = request.student_profile or {}
        student = StudentProfile(
            goal=profile_data.get("goal", "AI Engineer"),
            available_hours_per_week=float(profile_data.get("available_hours_per_week", 6.0))
        )

        # 2. Base path
        base_paths = path_service.generate_paths_for_student(student)
        if request.custom_base_path:
            original_path = request.custom_base_path
        else:
            # Default to Path A (Career Focus)
            original_path = base_paths[0]

        # Deep copy to construct alternative path
        alt_path = copy.deepcopy(original_path)
        scenario = request.scenario
        changes = []
        trade_offs = []

        op = scenario.operation.upper()

        if op == "REPLACE" or "replace" in scenario.operation.lower():
            target_name = (scenario.target or "AI Club").lower()
            replacement_type = scenario.replacement_type or "Research"

            # Find target step
            removed_step_name = ""
            new_steps = []
            for s in alt_path.steps:
                if target_name in s.name.lower() or target_name in s.type.lower():
                    removed_step_name = s.name
                    changes.append(f"Removed '{s.name}' ({s.hours_per_week}h/wk)")
                else:
                    new_steps.append(s)

            # If target step wasn't found specifically, replace the 3rd step (Club)
            if not removed_step_name and len(alt_path.steps) >= 3:
                removed_step_name = alt_path.steps[2].name
                changes.append(f"Removed '{removed_step_name}' ({alt_path.steps[2].hours_per_week}h/wk)")
                new_steps = [alt_path.steps[0], alt_path.steps[1], alt_path.steps[3]]

            # Add Research Replacement step
            research_step = PathStep(
                step_id="step_whatif_res_1",
                type="Research",
                name="RES_01: Autonomous Navigation & Perception Research",
                description="Faculty-led computer vision research with Prof. Aris Thorne.",
                hours_per_week=1.5,
                duration="1 Semester",
                skills=["Computer Vision", "Deep Learning", "Research Methods", "Faculty Mentorship"],
                reason="Replaces extracurricular club with direct faculty-led research exposure.",
                prerequisites=["AI101", "Python"],
                dependencies=["step_a_2"]
            )
            new_steps.append(research_step)
            changes.append(f"Added research alternative '{research_step.name}' ({research_step.hours_per_week}h/wk)")

            alt_path.steps = new_steps
            alt_path.title = "Path A' — What-If Alternative (Research Substituted)"
            alt_path.focus_type = "Research Substituted"
            alt_path.description = "Alternative candidate path substituting AI Club with faculty computer vision research."

            # Update metrics
            alt_path.metrics.research_exposure = min(100.0, original_path.metrics.research_exposure + 50.0)
            alt_path.metrics.faculty_interaction = min(100.0, original_path.metrics.faculty_interaction + 45.0)
            alt_path.metrics.networking_value = max(10.0, original_path.metrics.networking_value - 25.0)
            alt_path.metrics.industry_exposure = max(10.0, original_path.metrics.industry_exposure - 15.0)
            alt_path.metrics.goal_alignment = 95.0

            trade_offs.append("+ Research exposure increases from 40% to 90%")
            trade_offs.append("+ Faculty interaction increases from 45% to 90%")
            trade_offs.append("- Peer club networking decreases from 90% to 65%")
            trade_offs.append("✓ Fits within the exact same 6.0 hours/week limit")

        elif op in ("REDUCE_TIME", "4_HOURS") or scenario.new_time_limit is not None:
            new_limit = scenario.new_time_limit or 4.0
            alt_path.available_hours_per_week = new_limit
            
            # Trim steps to fit new limit
            trimmed_steps = []
            current_h = 0.0
            for s in alt_path.steps:
                if current_h + s.hours_per_week <= new_limit:
                    trimmed_steps.append(s)
                    current_h += s.hours_per_week
                else:
                    # Adjust hours if possible or skip
                    rem = new_limit - current_h
                    if rem >= 1.0:
                        s_copy = copy.deepcopy(s)
                        s_copy.hours_per_week = rem
                        s_copy.reason += f" (Reduced commitment to fit {new_limit}h limit)"
                        trimmed_steps.append(s_copy)
                        current_h += rem
                        changes.append(f"Reduced '{s.name}' commitment to {rem}h/wk")
                    else:
                        changes.append(f"Removed '{s.name}' to respect {new_limit}h/wk limit")

            alt_path.steps = trimmed_steps
            alt_path.title = f"Path A' — Optimized for {new_limit} Hours/Week"
            alt_path.description = f"Streamlined path tailored strictly for a {new_limit}-hour weekly commitment."

            alt_path.metrics.goal_alignment = 85.0
            alt_path.metrics.networking_value = max(10.0, original_path.metrics.networking_value - 20.0)

            trade_offs.append(f"✓ Total weekly time reduced from {original_path.total_hours_per_week}h to {new_limit}h")
            trade_offs.append(f"- Focused core steps retain 85% goal alignment")

        elif op == "REMOVE":
            target_name = (scenario.target or "Club").lower()
            new_steps = [s for s in alt_path.steps if target_name not in s.name.lower() and target_name not in s.type.lower()]
            alt_path.steps = new_steps
            changes.append(f"Removed target component '{scenario.target}'")
            trade_offs.append(f"✓ Reduced total time commitment to {sum(s.hours_per_week for s in new_steps)}h/wk")

        else:
            # Default generic adjustment
            changes.append(f"Applied scenario: {scenario.operation}")

        # Recalculate total hours
        alt_path.total_hours_per_week = sum(s.hours_per_week for s in alt_path.steps)
        alt_path.within_limit = (alt_path.total_hours_per_week <= alt_path.available_hours_per_week)

        # Unique skills gained
        all_skills = []
        for s in alt_path.steps:
            all_skills.extend(s.skills)
        alt_path.skills_gained = list(set(all_skills))

        # Generate metric comparisons
        m_comp = [
            MetricComparison(
                metric_name="Goal Alignment",
                original_val=original_path.metrics.goal_alignment,
                alternative_val=alt_path.metrics.goal_alignment,
                delta=round(alt_path.metrics.goal_alignment - original_path.metrics.goal_alignment, 1)
            ),
            MetricComparison(
                metric_name="Research Exposure",
                original_val=original_path.metrics.research_exposure,
                alternative_val=alt_path.metrics.research_exposure,
                delta=round(alt_path.metrics.research_exposure - original_path.metrics.research_exposure, 1)
            ),
            MetricComparison(
                metric_name="Networking Value",
                original_val=original_path.metrics.networking_value,
                alternative_val=alt_path.metrics.networking_value,
                delta=round(alt_path.metrics.networking_value - original_path.metrics.networking_value, 1)
            ),
            MetricComparison(
                metric_name="Faculty Interaction",
                original_val=original_path.metrics.faculty_interaction,
                alternative_val=alt_path.metrics.faculty_interaction,
                delta=round(alt_path.metrics.faculty_interaction - original_path.metrics.faculty_interaction, 1)
            ),
            MetricComparison(
                metric_name="Weekly Commitment (hrs)",
                original_val=original_path.total_hours_per_week,
                alternative_val=alt_path.total_hours_per_week,
                delta=round(alt_path.total_hours_per_week - original_path.total_hours_per_week, 1)
            )
        ]

        explanation = (
            f"What-If Transformation Complete: Substituted '{scenario.target or 'extracurriculars'}' "
            f"with {scenario.replacement_type or 'alternative opportunities'}. "
            f"The new path requires {alt_path.total_hours_per_week}h/week (limit: {alt_path.available_hours_per_week}h) "
            f"and shifts focus towards higher research and faculty mentorship."
        )

        return WhatIfResponse(
            original_path=original_path,
            alternative_path=alt_path,
            changes=changes,
            metric_comparisons=m_comp,
            trade_offs=trade_offs,
            explanation=explanation
        )

what_if_service = WhatIfService()
